### PR TITLE
Create Shooter Subsystem

### LINKED ISSUE #10: Create Shooter Subsystem
The shooter subsystem will contain two shooting motors and one "wrist" motor that will change the angle of the entire shooter assembly. I may decide to move the "wrist" motor into the elevator assembly due to the most likely larger interplay between those two actions. The shooter will have to be able to send a note game piece inside of its assembly, properly position it into "holding" and then be able to independently accelerate the non holding set of rollers. The holding rollers will push the note to the shooting rollers once they have gained goal roller velocity.

Some general requirements are characterization of all three motors. The wrist will need closed loop position control, while both rollers need to be controlled using closed loop velocity maintenance.

The assembly will contain a beam-break or some other type of sensor to be used in completing the above tasks and should also be made available to the rest of the robot project for note localization.

Once the above features are complete we will do an analysis of this subsystem in both function and relation to other mechanisms to come up with a state-machine.

### PR DESCRIPTION
**Pull request by coolinm**

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -6,15 +6,47 @@
...
+{
+  // Subsystems
+  private final Drivetrain m_Drivetrain = new Drivetrain(TunerConstants.DRIVETRAIN_CONSTANTS, TunerConstants.FRONT_LEFT,
+                      TunerConstants.FRONT_RIGHT, TunerConstants.BACK_LEFT, TunerConstants.BACK_RIGHT);
+  private final Shooter m_Shooter = new Shooter();
+  
+  // Controllers
+  private final CommandXboxController primaryController = new CommandXboxController(0);
+
+  public Drivetrain getDriveSub()
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/RobotContainer.java:30]: need public Shooter getShooter()

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -6,15 +6,47 @@
...
+  private void configureBindings() 
+  {
+    m_Drivetrain.setDefaultCommand(
+      new Drive(m_Drivetrain,
+                () -> -primaryController.getLeftY(),
+                () -> -primaryController.getLeftX(),
+                () -> -primaryController.getRightX()
+      ));
+
+    primaryController.a().onTrue(new SeedFieldRelative(m_Drivetrain));
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/RobotContainer.java]: get this out of here

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -6,15 +6,47 @@
...
+public class RobotContainer 
+{
+  // Subsystems
+  private final Drivetrain m_Drivetrain = new Drivetrain(TunerConstants.DRIVETRAIN_CONSTANTS, TunerConstants.FRONT_LEFT,
+                      TunerConstants.FRONT_RIGHT, TunerConstants.BACK_LEFT, TunerConstants.BACK_RIGHT);
+  private final Shooter m_Shooter = new Shooter();
+  
+  // Controllers
+  private final CommandXboxController primaryController = new CommandXboxController(0);
+
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/RobotContainer.java:25]: move constructor here

File: src/main/java/frc/robot/commands/shooter/SetShooterMotor1Percent.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Percent = percent;
+    addRequirements(shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize()
+  {
+    m_Shooter.setShooterMotor1Percent(m_Percent);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetShooterMotor1Percent.java]: this.m_Percent

File: src/main/java/frc/robot/commands/shooter/SetShooterMotor1Velocity.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Velocity = velocity;
+    addRequirements(shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize()
+  {
+     m_Shooter.setMotor1Velocity(m_Velocity);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetShooterMotor1Velocity.java]: this

File: src/main/java/frc/robot/commands/shooter/SetShooterMotor2Percent.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Percent = percent;
+    addRequirements(shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize()
+  {
+    m_Shooter.setShooterMotor2Percent(m_Percent);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetShooterMotor2Percent.java]: this

File: src/main/java/frc/robot/commands/shooter/SetShooterMotor2Velocity.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Velocity = velocity;
+    addRequirements(shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() 
+  {
+    m_Shooter.setMotor2Velocity(m_Velocity);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetShooterMotor2Velocity.java]: this

File: src/main/java/frc/robot/commands/shooter/SetWristMotorPercent.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Percent = percent;
+    addRequirements(m_Shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() 
+  {
+    m_Shooter.setWristPercent(m_Percent);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetWristMotorPercent.java]: this

File: src/main/java/frc/robot/commands/shooter/SetWristMotorPosition.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    this.m_Shooter = shooter;
+    this.m_Position = position;
+    addRequirements(m_Shooter);
+  }
+
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() 
+  {
+    m_Shooter.setWristPos(m_Position);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/shooter/SetWristMotorPosition.java]: this

File: src/main/java/frc/robot/constants/shooter/ShooterConstants.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package frc.robot.constants.shooter;
+
+import com.ctre.phoenix6.configs.Slot0Configs;
+
+public class ShooterConstants
+{
+    // Gains for the shooter motors
+    public static final Slot0Configs SHOOTER_GAINS = new Slot0Configs()
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/constants/shooter/ShooterConstants.java]: add another gains

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  private final TalonFX m_ShooterMotor2 = new TalonFX(Hardware.SHOOTER_MOTOR_2_ID);
+  private final TalonFX m_WristMotor = new TalonFX(Hardware.WRIST_MOTOR_ID);
+
+    /**
+   * @brief Creates a new Shooter subsystem.
+  */
+  public Shooter() 
+  {
+    m_ShooterMotor1.getConfigurator().apply(ShooterConstants.SHOOTER_GAINS);
+    m_ShooterMotor2.getConfigurator().apply(ShooterConstants.SHOOTER_GAINS);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Shooter.java]: apply after you create it

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  /**
+   * @brief Applies a Motion Magic request to the selected motor with the set velocity.
+   * @param motor The motor you want to send the request to MOTOR_1 or MOTOR_2, setting this to WRIST_MOTOR will do nothing.
+   * @param velocity The desired velocity to run the motor at, measured in rotations per second.
+  */
+  public void setMotorVelocity(ShooterConstants.MOTOR motor, double velocity)
+  {
+    MotionMagicVelocityVoltage request = new MotionMagicVelocityVoltage(velocity, 40, false, 0, 0, false, false, false);
+
+    switch (motor){
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Shooter.java]: bad opening bracker

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+  // Shooter Functions
+
+  /**
+   * @brief Applies a Motion Magic request to the selected motor with the set velocity.
+   * @param motor The motor you want to send the request to MOTOR_1 or MOTOR_2, setting this to WRIST_MOTOR will do nothing.
+   * @param velocity The desired velocity to run the motor at, measured in rotations per second.
+  */
+  public void setMotorVelocity(ShooterConstants.MOTOR motor, double velocity)
+  {
+    MotionMagicVelocityVoltage request = new MotionMagicVelocityVoltage(velocity, 40, false, 0, 0, false, false, false);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Shooter.java]: make this 40 a constant

**Review by jasonsoohoo — CHANGES_REQUESTED** [review]
needs changes

**Review by jasonsoohoo — APPROVED** [review]

**Review by cjahner1 — APPROVED** [review]

**Pull request merged**